zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,505 @@
+/// hash from just the personalization string.
+#[cfg(feature = "zfuture")]
+pub(crate) fn tze_outputs_hash(tze_outputs: &[TzeOut]) -> Blake2bHash {
+    let mut h = HashWriter::new(ZCASH_TZE_OUTPUTS_HASH_PERSONALIZATION);
+    for tzeout in tze_outputs {
+        tzeout.write(&mut h).unwrap();
+    }
+    h.finalize()
+}
+
+/// Write disjoint parts of each spend  to a pair of hashes:
+/// * [nullifier*] - personalized with ZCASH_SAPLING_SPENDS_COMPACT_HASH_PERSONALIZATION
+/// * [(cv, anchor, rk, zkproof)*] - personalized with ZCASH_SAPLING_SPENDS_NONCOMPACT_HASH_PERSONALIZATION
+///
+/// Then, hash these together personalized by ZCASH_SAPLING_SPENDS_HASH_PERSONALIZATION
+pub(crate) fn shielded_spends_hash<A: sapling::Authorization>(
+    shielded_spends: &[SpendDescription<A>],
+) -> Blake2bHash {
+    let mut ch = HashWriter::new(ZCASH_SAPLING_SPENDS_COMPACT_HASH_PERSONALIZATION);
+    let mut nh = HashWriter::new(ZCASH_SAPLING_SPENDS_NONCOMPACT_HASH_PERSONALIZATION);
+    for s_spend in shielded_spends {
+        // we build the hash of nullifiers separately for compact blocks.
+        ch.write_all(&s_spend.nullifier.as_ref()).unwrap();
+
+        nh.write_all(&s_spend.cv.to_bytes()).unwrap();
+        nh.write_all(&s_spend.anchor.to_repr()).unwrap();
+        s_spend.rk.write(&mut nh).unwrap();
+    }
+
+    let mut h = HashWriter::new(ZCASH_SAPLING_SPENDS_HASH_PERSONALIZATION);
+    h.write_all(&ch.finalize().as_bytes()).unwrap();
+    h.write_all(&nh.finalize().as_bytes()).unwrap();
+    h.finalize()
+}
+
+/// Write disjoint parts of each output as 3 separate hashes:
+/// * [(cmu, epk, enc_ciphertext[..52])*] personalized with ZCASH_SAPLING_OUTPUTS_COMPACT_HASH_PERSONALIZATION
+/// * [enc_ciphertext[52..564]*] (memo ciphertexts) personalized with ZCASH_SAPLING_OUTPUTS_MEMOS_HASH_PERSONALIZATION
+/// * [(cv, enc_ciphertext[564..], out_ciphertext, zkproof)*] personalized with ZCASH_SAPLING_OUTPUTS_NONCOMPACT_HASH_PERSONALIZATION
+///
+/// Then, hash these together personalized with ZCASH_SAPLING_OUTPUTS_HASH_PERSONALIZATION
+pub(crate) fn shielded_outputs_hash(shielded_outputs: &[OutputDescription]) -> Blake2bHash {
+    let mut ch = HashWriter::new(ZCASH_SAPLING_OUTPUTS_COMPACT_HASH_PERSONALIZATION);
+    let mut mh = HashWriter::new(ZCASH_SAPLING_OUTPUTS_MEMOS_HASH_PERSONALIZATION);
+    let mut nh = HashWriter::new(ZCASH_SAPLING_OUTPUTS_NONCOMPACT_HASH_PERSONALIZATION);
+    for s_out in shielded_outputs {
+        ch.write_all(&s_out.cmu.to_repr().as_ref()).unwrap();
+        ch.write_all(&s_out.ephemeral_key.to_bytes()).unwrap();
+        ch.write_all(&s_out.enc_ciphertext[..52]).unwrap();
+
+        mh.write_all(&s_out.enc_ciphertext[52..564]).unwrap();
+
+        nh.write_all(&s_out.cv.to_bytes()).unwrap();
+        nh.write_all(&s_out.enc_ciphertext[564..]).unwrap();
+        nh.write_all(&s_out.out_ciphertext).unwrap();
+        nh.write_all(&s_out.zkproof).unwrap();
```

Comment:
Sorry, forgot to comment about this one:

```suggestion
```

I think this is not needed. ZIP-244 says "excluding zkproof data", also https://github.com/zcash-hackworks/zcash-test-vectors/pull/20 does not include it